From 6b146ce4e3ff3e34f722547ed81171b5eff343fb Mon Sep 17 00:00:00 2001
From: David Stinemetze <davidstinemetze@gmail.com>
Date: Tue, 12 Mar 2019 23:57:15 -0500
Subject: [PATCH] 3039408
diff --git a/core/composer.json b/core/composer.json
index 34fed39b1a..bbd9faa42e 100644
--- a/core/composer.json
+++ b/core/composer.json
@@ -32,7 +32,7 @@
         "symfony/polyfill-iconv": "^1.0",
         "symfony/yaml": "~3.4.5",
         "typo3/phar-stream-wrapper": "^2.0.1",
-        "twig/twig": "^1.35.0",
+        "twig/twig": "^1.38.2",
         "doctrine/common": "^2.5",
         "doctrine/annotations": "^1.2",
         "guzzlehttp/guzzle": "^6.2.1",
diff --git a/core/lib/Drupal/Core/Template/TwigNodeTrans.php b/core/lib/Drupal/Core/Template/TwigNodeTrans.php
index c0691dbca6..aad9d0b069 100644
--- a/core/lib/Drupal/Core/Template/TwigNodeTrans.php
+++ b/core/lib/Drupal/Core/Template/TwigNodeTrans.php
@@ -2,6 +2,8 @@
 
 namespace Drupal\Core\Template;
 
+use Twig\Node\CheckToStringNode;
+
 /**
  * A class that defines the Twig 'trans' tag for Drupal.
  *
@@ -113,6 +115,9 @@ protected function compileString(\Twig_Node $body) {
             $n = $n->getNode('node');
           }
 
+          if ($n instanceof CheckToStringNode) {
+            $n = $n->getNode('expr');
+          }
           $args = $n;
 
           // Support TwigExtension->renderVar() function in chain.
@@ -134,6 +139,9 @@ protected function compileString(\Twig_Node $body) {
             }
             $args = $args->getNode('node');
           }
+          if ($args instanceof CheckToStringNode) {
+            $args = $args->getNode('expr');
+          }
           if ($args instanceof \Twig_Node_Expression_GetAttr) {
             $argName = [];
             // Reuse the incoming expression.
diff --git a/core/modules/system/tests/modules/twig_extension_test/src/TwigExtension/TestExtension.php b/core/modules/system/tests/modules/twig_extension_test/src/TwigExtension/TestExtension.php
index 93be201e5c..2f51e721a6 100644
--- a/core/modules/system/tests/modules/twig_extension_test/src/TwigExtension/TestExtension.php
+++ b/core/modules/system/tests/modules/twig_extension_test/src/TwigExtension/TestExtension.php
@@ -2,6 +2,9 @@
 
 namespace Drupal\twig_extension_test\TwigExtension;
 
+use Twig\TwigFilter;
+use Twig\TwigFunction;
+
 /**
  * A test Twig extension that adds a custom function and a custom filter.
  */
@@ -21,7 +24,7 @@ class TestExtension extends \Twig_Extension {
    */
   public function getFunctions() {
     return [
-      'testfunc' => new \Twig_Function_Function(['Drupal\twig_extension_test\TwigExtension\TestExtension', 'testFunction']),
+      'testfunc' => new TwigFunction('testfunc', ['Drupal\twig_extension_test\TwigExtension\TestExtension', 'testFunction']),
     ];
   }
 
@@ -39,7 +42,7 @@ public function getFunctions() {
    */
   public function getFilters() {
     return [
-      'testfilter' => new \Twig_Filter_Function(['Drupal\twig_extension_test\TwigExtension\TestExtension', 'testFilter']),
+      'testfilter' => new TwigFilter('testfilter', ['Drupal\twig_extension_test\TwigExtension\TestExtension', 'testFilter']),
     ];
   }
 
diff --git a/core/modules/system/tests/src/Functional/Theme/TwigRegistryLoaderTest.php b/core/modules/system/tests/src/Functional/Theme/TwigRegistryLoaderTest.php
index 2446289984..da02b2ba8a 100644
--- a/core/modules/system/tests/src/Functional/Theme/TwigRegistryLoaderTest.php
+++ b/core/modules/system/tests/src/Functional/Theme/TwigRegistryLoaderTest.php
@@ -3,6 +3,7 @@
 namespace Drupal\Tests\system\Functional\Theme;
 
 use Drupal\Tests\BrowserTestBase;
+use Twig\TemplateWrapper;
 
 /**
  * Tests Twig registry loader.
@@ -33,7 +34,7 @@ protected function setUp() {
    * Checks to see if a value is a Twig template.
    */
   public function assertTwigTemplate($value, $message = '', $group = 'Other') {
-    $this->assertTrue($value instanceof \Twig_Template, $message, $group);
+    $this->assertTrue($value instanceof TemplateWrapper, $message, $group);
   }
 
   /**
diff --git a/core/modules/system/tests/src/Functional/Theme/TwigSettingsTest.php b/core/modules/system/tests/src/Functional/Theme/TwigSettingsTest.php
index 0ad88ce320..0aed306fd9 100644
--- a/core/modules/system/tests/src/Functional/Theme/TwigSettingsTest.php
+++ b/core/modules/system/tests/src/Functional/Theme/TwigSettingsTest.php
@@ -96,7 +96,12 @@ public function testTwigCacheOverride() {
     // theme_test.template_test.html.twig.
     $info = $templates->get('theme_test_template_test');
     $template_filename = $info['path'] . '/' . $info['template'] . $extension;
-    $cache_filename = $this->container->get('twig')->getCacheFilename($template_filename);
+
+    /** @var \Drupal\Core\Template\TwigEnvironment $twig */
+    $environment = $this->container->get('twig');
+    $cache = $environment->getCache();
+    $class = $environment->getTemplateClass($template_filename);
+    $cache_filename = $cache->generateKey($template_filename, $class);
 
     // Navigate to the page and make sure the template gets cached.
     $this->drupalGet('theme-test/template-test');
diff --git a/core/modules/system/tests/src/Kernel/Theme/TwigNamespaceTest.php b/core/modules/system/tests/src/Kernel/Theme/TwigNamespaceTest.php
index 4ffd2558f0..87a9801ec1 100644
--- a/core/modules/system/tests/src/Kernel/Theme/TwigNamespaceTest.php
+++ b/core/modules/system/tests/src/Kernel/Theme/TwigNamespaceTest.php
@@ -3,6 +3,7 @@
 namespace Drupal\Tests\system\Kernel\Theme;
 
 use Drupal\KernelTests\KernelTestBase;
+use Twig\TemplateWrapper;
 
 /**
  * Tests Twig namespaces.
@@ -33,7 +34,7 @@ protected function setUp() {
    * Checks to see if a value is a twig template.
    */
   public function assertTwigTemplate($value, $message = '', $group = 'Other') {
-    $this->assertTrue($value instanceof \Twig_Template, $message, $group);
+    $this->assertTrue($value instanceof TemplateWrapper, $message, $group);
   }
 
   /**
diff --git a/core/tests/Drupal/KernelTests/Core/Theme/TwigEnvironmentTest.php b/core/tests/Drupal/KernelTests/Core/Theme/TwigEnvironmentTest.php
index c95b0545b9..6ff2f46304 100644
--- a/core/tests/Drupal/KernelTests/Core/Theme/TwigEnvironmentTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Theme/TwigEnvironmentTest.php
@@ -163,12 +163,19 @@ public function testCacheFilename() {
     $expected = strlen($prefix) + 2 + 2 * TwigPhpStorageCache::SUFFIX_SUBSTRING_LENGTH;
     $this->assertEquals($expected, strlen($key));
 
-    $original_filename = $environment->getCacheFilename('core/modules/system/templates/container.html.twig');
+    $template_filename = 'core/modules/system/templates/container.html.twig';
+    $cache = $environment->getCache();
+    $class = $environment->getTemplateClass($template_filename);
+    $original_filename = $cache->generateKey($template_filename, $class);
+
     \Drupal::getContainer()->set('twig', NULL);
 
     \Drupal::service('module_installer')->install(['twig_extension_test']);
     $environment = \Drupal::service('twig');
-    $new_extension_filename = $environment->getCacheFilename('core/modules/system/templates/container.html.twig');
+    $cache = $environment->getCache();
+    $class = $environment->getTemplateClass($template_filename);
+    $new_extension_filename = $cache->generateKey($template_filename, $class);
+
     \Drupal::getContainer()->set('twig', NULL);
 
     $this->assertNotEqual($new_extension_filename, $original_filename);
-- 
2.15.1 (Apple Git-101)

